batcher/sweep_batch.go

```diff
@@ -0,0 +1,1227 @@
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+const (
+	defaultFeeRateStep     = chainfee.SatPerKWeight(100)
+	defaultBatchConfTarget = 12
+	batchConfHeight        = 3
+	batchPublishDelay      = time.Millisecond * 100
+)
+
+// Sweep stores any data related to sweeping a specific outpoint.
+type Sweep struct {
+	// SwapHash is the hash of the swap that the sweep belongs to.
+	SwapHash lntypes.Hash
+
+	// Outpoint is the outpoint being swept.
+	Outpoint wire.OutPoint
+
+	// Value is the value of the outpoint being swept.
+	Value btcutil.Amount
+
+	// ConfTarget is the confirmation target of the sweep.
+	ConfTarget int32
+
+	// Timeout is the timeout of the swap that the sweep belongs to.
+	Timeout int32
+
+	// InitiationHeight is the height at which the swap was initiated.
+	InitiationHeight int32
+
+	// Htlc is the HTLC that is being swept.
+	Htlc swap.Htlc
+
+	// Preimage is the preimage of the HTLC that is being swept.
+	Preimage lntypes.Preimage
+
+	// SwapInvoicePaymentAddr is the payment address of the swap invoice.
+	SwapInvoicePaymentAddr [32]byte
+
+	// Key is the key used to sign the HTLC.
+	Key [33]byte
+
+	// HtlcKeys is the set of keys used to sign the HTLC.
+	HtlcKeys loopdb.HtlcKeys
+
+	// HtlcSuccessEstimator is a function that estimates the weight of the
+	// HTLC success script.
+	HtlcSuccessEstimator func(*input.TxWeightEstimator) error
+
+	// ProtocolVersion is the protocol version of the swap that the sweep
+	// belongs to.
+	ProtocolVersion loopdb.ProtocolVersion
+
+	// IsWalletAddr is true if the sweep spends to a wallet address.
+	IsWalletAddr bool
+
+	// DestAddr is the destination address of the sweep.
+	DestAddr btcutil.Address
+}
+
+// batchState is the state of the batch.
+type batchState uint8
+
+const (
+	// Open is the state in which the batch is able to accept new sweeps.
+	Open batchState = iota
+
+	// Closed is the state in which the batch is no longer able to accept
+	// new sweeps.
+	Closed
+
+	// Confirmed is the state in which the batch transaction has reached the
+	// configured conf height.
+	Confirmed
+)
+
+// BatchConfig is the configuration for a batch.
+type BatchConfig struct {
+	// MaxTimeoutDistance is the maximum timeout distance that 2 distinct
+	// sweeps can have in the same batch.
+	MaxTimeoutDistance int32
+
+	// BatchConfTarget is the confirmation target of the batch transaction.
+	BatchConfTarget int32
+}
+
+// RBFCache stores data related to our last fee bump.
+type RBFCache struct {
+	// LastHeight is the last height at which we increased our feerate.
+	LastHeight int32
+
+	// FeeRate is the last used fee rate we used to publish a batch tx.
+	FeeRate chainfee.SatPerKWeight
+}
+
+// Batch is a collection of sweeps that are published together.
+type Batch struct {
+	// ID is the primary identifier of this batch.
+	ID int32
+
+	// State is the current State of the batch.
+	State batchState
+
+	// PrimarySweepID is the swap hash of the primary sweep in the batch.
```

Comment:
Sure:
The primary sweep refers to the first sweep that entered the batch, which effectively is the sweep that created this batch. The reason we keep track of it is because, being the very first sweep in the batch, it is always present in all the published batch transactions. That allows us to only register a single spend monitor for every batch (which monitors the primary sweep's outpoint) and successfully retrieve the batch tx that got confirmed.